src/H5Dio.c

```diff
@@ -810,98 +810,35 @@ H5D__ioinfo_adjust(H5D_io_info_t *io_info, const H5D_t *dset, const H5S_t *file_
-                         */
-                        if (local_buffer_space && local_error_message_previously_written) {
-                            HDstrncat(local_no_collective_cause_string, "; ", local_buffer_space);
-                            local_buffer_space -= MIN(local_buffer_space, 2);
-                        }
-
-                        if (local_buffer_space)
-                            HDstrncat(local_no_collective_cause_string, cause_strings[idx],
-                                      local_buffer_space);
-
-                        local_error_message_previously_written = TRUE;
-                    } /* end if */
-
-                    if (cause & global_no_collective_cause) {
-                        size_t global_buffer_space = sizeof(global_no_collective_cause_string) -
-                                                     HDstrlen(global_no_collective_cause_string) - 1;
-
-                        /* Check if there were any previous error messages included. If so, prepend a
-                         * semicolon to separate the messages.
-                         */
-                        if (global_buffer_space && global_error_message_previously_written) {
-                            HDstrncat(global_no_collective_cause_string, "; ", global_buffer_space);
-                            global_buffer_space -= MIN(global_buffer_space, 2);
-                        }
-
-                        if (global_buffer_space)
-                            HDstrncat(global_no_collective_cause_string, cause_strings[idx],
-                                      global_buffer_space);
-
-                        global_error_message_previously_written = TRUE;
-                    } /* end if */
-                }     /* end for */
-
-                HGOTO_ERROR(H5E_IO, H5E_NO_INDEPENDENT, FAIL,
-                            "Can't perform independent write with filters in pipeline.\n"
-                            "    The following caused a break from collective I/O:\n"
-                            "        Local causes: %s\n"
-                            "        Global causes: %s",
-                            local_no_collective_cause_string, global_no_collective_cause_string);
-            } /* end if */
+            if (io_info->op_type == H5D_IO_OP_WRITE && io_info->dset->shared->dcpl_cache.pline.nused > 0) {
+                int comm_size = 0;
+
+                /* Retrieve size of MPI communicator used for file */
+                if ((comm_size = H5F_shared_mpi_get_size(io_info->f_sh)) < 0)
+                    HGOTO_ERROR(H5E_FILE, H5E_CANTGET, FAIL, "can't get MPI communicator size")
+
+                if (comm_size > 1) {
+                    char local_no_coll_cause_string[512];
+                    char global_no_coll_cause_string[512];
+
+                    if (H5D__mpio_get_no_coll_cause_strings(local_no_coll_cause_string, 512,
+                                                            global_no_coll_cause_string, 512) < 0)
```

Comment:
Moved most of this code into a new function to get strings for the reasons why collective I/O was broken for code reuse.